### PR TITLE
Account Virtual Addresses

### PR DESCRIPTION
**Pull request by Nana-EC**
Signed-off-by: Nana Essilfie-Conduah <nana@swirldslabs.com>

**Description**:
Account Virtual Addresses support to support EVM equivalence

The alias concept from [HIP 32](https://hips.hedera.com/hip/hip-32) was modified to help alleviate the challenges of account EVM identification by providing a source for the EVM address an account could adopt for EVM compatibility. However, this had its limitations as it was not originally intended to encompass EVM address logic which is fundamentally based on `ECDSA` keys and calculations on accounts state, not the next available entity identifier.

Thus, the proposed solution is to allow an account to contain 1 or more `evmAddress` values for ECDSA keys it owns that dictate how an account appears on the EVM. These addresses will be known as Virtual Addresses.

**Related issue(s)**:

Fixes #

**Notes for reviewer**:
This HIP builds upon HIP 583 and will be followed by an additional HIP for system contract support of virtual addresses

**Checklist**

- [ ] Documented (Code comments, README, etc.)
- [ ] Tested (unit, integration, etc.)

### CODE REVIEW

File: HIP/hip-0000-virtual-addresses.md [hunk 1]
```diff
@@ -0,0 +1,543 @@
...
+This is contrary to the EVM view where an account encompasses an account identifier which is a derivative of its static cryptographic public signing key i.e. public address (as noted in [HIP 583](https://hips.hedera.com/hip/hip-583)). Here, it is not possible to dissociate an account from its public key as you can in Hedera. 
+
+This has made it challenging when a Hedera account needs to be represented on the EVM or when an EVM feature that relies on the public key to account identifier mapping needs to be honored by the Hedera network. The alias concept from [HIP 32](https://hips.hedera.com/hip/hip-32) was modified to help alleviate this issue by providing a source for the EVM address an account would adopt on EVM transactions. However, this had its limitations as it was not originally intended to encompass EVM address logic which is fundamentally based on `ECDSA` keys and calculations on accounts state, not the next available entity identifier.
```
[the remark this answers is not in the document]
Comment (Nana-EC) [reply HIP/hip-0000-virtual-addresses.md]: Addressed

File: HIP/hip-0000-virtual-addresses.md [hunk 2]
```diff
@@ -0,0 +1,543 @@
...
+
+For easy reference new and clarifying terms are listed here
+
+- `ETH contracts`: Contracts lifted from Ethereum or other EVM compatible ledgers in which a maintainer does not desire to change code in porting and or explicitly wants to use precompile contract operations such as `ECRECOVER`.
+
+- `Hedera contracts`: Smart Contracts designed with additional logic to utilize Hedera system smart contracts.
+
+- `Hedera address` - Also known as the `long-zero` address (as noted in [HIP 583](https://hips.hedera.com/hip/hip-583)) account identifier, this is the result of encoding the `<shard>.<relam>.<num>` integer values into a 20 bytes hex format to be conformant with ECDSA public addresses. This is a more explicit naming for exposure in products that highlights the essence of the address.
+
+- Alias - An account alias can be an `ECDSA`/`ED25519` public key (per HIP 32). Notably, in the past the alias was sometimes used to refer to an `ECDSA` public address, however, it should solely refer to a public key.
```
[the remark this answers is not in the document]
Comment (Nana-EC) [reply HIP/hip-0000-virtual-addresses.md]: Addressed

File: HIP/hip-0000-virtual-addresses.md [hunk 3]
```diff
@@ -0,0 +1,543 @@
...
+
+This inadvertently presents multiple points of Hedera vs Ethereum design choices that have occasionally reduced user experience and resulted in complicated scenarios.
+
+For easy reference new and clarifying terms are listed here
+
+- `ETH contracts`: Contracts lifted from Ethereum or other EVM compatible ledgers in which a maintainer does not desire to change code in porting and or explicitly wants to use precompile contract operations such as `ECRECOVER`.
+
+- `Hedera contracts`: Smart Contracts designed with additional logic to utilize Hedera system smart contracts.
+
+- `Hedera address` - Also known as the `long-zero` address (as noted in [HIP 583](https://hips.hedera.com/hip/hip-583)) account identifier, this is the result of encoding the `<shard>.<relam>.<num>` integer values into a 20 bytes hex format to be conformant with ECDSA public addresses. This is a more explicit naming for exposure in products that highlights the essence of the address.
```
[the remark this answers is not in the document]
Comment (Nana-EC) [reply HIP/hip-0000-virtual-addresses.md]: Addressed

File: HIP/hip-0000-virtual-addresses.md [hunk 4]
```diff
@@ -0,0 +1,543 @@
...
+
+The ledger does this while maintaining and advancing Hedera API (HAPI) transactions that provide native support for Account, Contract, File, Topic, Token and Schedule entities.
+
+This inadvertently presents multiple points of Hedera vs Ethereum design choices that have occasionally reduced user experience and resulted in complicated scenarios.
+
+For easy reference new and clarifying terms are listed here
+
+- `ETH contracts`: Contracts lifted from Ethereum or other EVM compatible ledgers in which a maintainer does not desire to change code in porting and or explicitly wants to use precompile contract operations such as `ECRECOVER`.
+
+- `Hedera contracts`: Smart Contracts designed with additional logic to utilize Hedera system smart contracts.
```
[the remark this answers is not in the document]
Comment (Nana-EC) [reply HIP/hip-0000-virtual-addresses.md]: Updated

File: HIP/hip-0000-virtual-addresses.md [hunk 5]
```diff
@@ -0,0 +1,543 @@
...
+
+The Hedera Smart Contract Service (HSCS) processes EVM aimed transactions in a fast and cheap manner, with logic that is equivalent to Ethereum utiliing theesu EVM client. 
+
+The ledger does this while maintaining and advancing Hedera API (HAPI) transactions that provide native support for Account, Contract, File, Topic, Token and Schedule entities.
+
+This inadvertently presents multiple points of Hedera vs Ethereum design choices that have occasionally reduced user experience and resulted in complicated scenarios.
+
+For easy reference new and clarifying terms are listed here
+
+- `ETH contracts`: Contracts lifted from Ethereum or other EVM compatible ledgers in which a maintainer does not desire to change code in porting and or explicitly wants to use precompile contract operations such as `ECRECOVER`.
```
[the remark this answers is not in the document]
Comment (Nana-EC) [reply HIP/hip-0000-virtual-addresses.md]: Fixed

File: HIP/hip-0000-virtual-addresses.md [hunk 6]
```diff
@@ -0,0 +1,543 @@
...
+
+## Abstract
+
+The Hedera Smart Contract Service (HSCS) processes EVM aimed transactions in a fast and cheap manner, with logic that is equivalent to Ethereum utiliing theesu EVM client. 
+
+The ledger does this while maintaining and advancing Hedera API (HAPI) transactions that provide native support for Account, Contract, File, Topic, Token and Schedule entities.
+
+This inadvertently presents multiple points of Hedera vs Ethereum design choices that have occasionally reduced user experience and resulted in complicated scenarios.
+
+For easy reference new and clarifying terms are listed here
```
[the remark this answers is not in the document]
Comment (Nana-EC) [reply HIP/hip-0000-virtual-addresses.md]: Fixed

File: HIP/hip-0000-virtual-addresses.md [hunk 7]
```diff
@@ -0,0 +1,543 @@
...
+superseded-by: 
+---
+
+## Abstract
+
+The Hedera Smart Contract Service (HSCS) processes EVM aimed transactions in a fast and cheap manner, with logic that is equivalent to Ethereum utiliing theesu EVM client. 
+
+The ledger does this while maintaining and advancing Hedera API (HAPI) transactions that provide native support for Account, Contract, File, Topic, Token and Schedule entities.
+
+This inadvertently presents multiple points of Hedera vs Ethereum design choices that have occasionally reduced user experience and resulted in complicated scenarios.
```
[the remark this answers is not in the document]
Comment (Nana-EC) [reply HIP/hip-0000-virtual-addresses.md]: Updated

**Comment by se7enarianelabs** [comment]
Will the virtual addresses influence the jRPC relay?

**Comment by Nana-EC** [comment]
> Will the virtual addresses influence the jRPC relay?

Thanks @se7enarianelabs this will not directly impact the relay. No changes will be necessary there.

**Comment by se7enarianelabs** [comment]
> > Will the virtual addresses influence the jRPC relay?
> 
> Thanks @se7enarianelabs this will not directly impact the relay. No changes will be necessary there.

Okay, thanks for the clarification. I was thinking about a case where two different Hedera accounts share the same virtual address.

File: HIP/hip-631.md [hunk 8]
```diff
@@ -0,0 +1,555 @@
...
+This should be with no little to no overhead for users, minimal efforts to devs and with no possibility of loss of assets in any migrations.
+
+## Rationale
+
+On Hedera a complete and functioning account can be identified at a point in time by 
+- an account identifier `shard.realm.num` (dictated by the next available entity counter)
+- a modifiable cryptographic public signing key
+- an optional static public key account identifier i.e. alias.
+
+This differes from the EVM view where an account can always be identified by a single public address (as noted in [HIP 583](https://hips.hedera.com/hip/hip-583)) derived from an ECDSA cryptographic public signing key. Here, it is not possible to dissociate an account from its public key as you can in Hedera.
```
Comment (MiroslavGatsanoga) [inline HIP/hip-631.md:65]: **typo**: `differes` -> `differs`

File: HIP/hip-631.md [hunk 9]
```diff
@@ -0,0 +1,555 @@
...
+- an account identifier `shard.realm.num` (dictated by the next available entity counter)
+- a modifiable cryptographic public signing key
+- an optional static public key account identifier i.e. alias.
+
+This differes from the EVM view where an account can always be identified by a single public address (as noted in [HIP 583](https://hips.hedera.com/hip/hip-583)) derived from an ECDSA cryptographic public signing key. Here, it is not possible to dissociate an account from its public key as you can in Hedera. 
+
+This has made it challenging when a Hedera account needs to be represented on the EVM or when an EVM operation that relies on the public key to public address mapping is expected on the Hedera network. The alias concept from [HIP 32](https://hips.hedera.com/hip/hip-32) was modified to help alleviate this issue by providing a source for the EVM address that an account would adopt whene executing EVM transactions. However, this had its limitations as it was not originally intended to encompass EVM address logic which is fundamentally based on `ECDSA` keys and calculations of EVM account state.
```
Comment (MiroslavGatsanoga) [inline HIP/hip-631.md:67]: **typo**: `whene` -> `when`

File: HIP/hip-631.md [hunk 10]
```diff
@@ -0,0 +1,555 @@
...
+        *
+        * If a transaction lazily-creates this account, a subsequent transaction will be required containing the public key bytes
+        * that map to the EVM address bytes. Lazy account creates will only support the keccak-256 hash of a ECDSA_SECP256K1 primitive key form.
+        */
+        bytes evm_address = 5;
+    }
+}
+```
+
+Expose the virtual addresses in a `CrytptoGetInfo` response by updating `AccountInfo`
```
Comment (MiroslavGatsanoga) [inline HIP/hip-631.md:164]: **typo**: `CrytptoGetInfo` -> `CryptoGetInfo`

File: HIP/hip-631.md [hunk 11]
```diff
@@ -0,0 +1,555 @@
...
+
+Note2: Account fee logic will incorporate the extra virtual address property to be stored in state.
+
+### Support virtual address list modification
+
+The list of virtual addresses should support the ability to add, disable and remove entries.
+
+To add a virtual address an account would submit a `CryptoUpdateTransactionBody` with a `virtual_address_update.add` containing the desired evmAddress. The transaction must be signed by the ECDSA private key that maps to the evm address being added.
+
+To disable a virtual address an account would submit a `CryptoUpdateTransactionBody` with a `virtual_address_update.disable` containing the existing evmAddress. The transaction must be signed by the ECDSA private key that maps to the evm address being added.
```
Comment (MiroslavGatsanoga) [inline HIP/hip-631.md:341]: **typo**: `being added` -> `being disabled`

File: HIP/hip-631.md [hunk 12]
```diff
@@ -0,0 +1,555 @@
...
+
+To remove a virtual address an account would submit a `CryptoUpdateTransactionBody` with a `virtual_address_update.remove` containing the existing evmAddress. The transaction must be signed by the ECDSA private key that maps to the evm address being added.
```
Comment (MiroslavGatsanoga) [inline HIP/hip-631.md:343]: **typo**: `being added` -> `being removed`

File: HIP/hip-631.md [hunk 13]
```diff
@@ -0,0 +1,555 @@
...
+|  | key | N | evm_address = new | normal create non-existing account with new virtual address. The provided virtual address will serve as the default virtual address for the account. |
+|  | evm_address | N |  | lazy-create non-existing account with new virtual address. The provided virtual address will serve as the default virtual address for the account. |
+|  |  |  |  |  |
+| CryptoTransfer | accountId.alias | N | evm_address = new | transfer to non-existing account (auto-create with new virtual address). The provided virtual address will serve as the default virtual address for the account. |
+|  | accountId.alias | Y | - | transfer to existing |
+|  | accountId.evmAddress | N | evm_address = new | transfer to non-existing account (lazy-create with new virtual address). The provided virtual address will serve as the default virtual address for the account. |
+|  | accountId.evmAddress | Y | - | transfer to existing |
+|  |  |  |  |  |
+| CryptoUpdate | virtual_address_update.add.evmAddress = new (optional) virtualAddress.isDefault = true | Y | - | Add virtual address to existing account. Pull details from transaction body (not populating record preserves space) |
```
Comment (MiroslavGatsanoga) [inline HIP/hip-631.md:266]: We should have a row for disable as well.

File: HIP/hip-631.md [hunk 14]
```diff
@@ -0,0 +1,555 @@
...
+
+With the addition of virtual addresses and the separation of alias, network logic allows a contract account to represent its contract address with its virtual address.
+
+## Community product impacts
+
+### SDK Impacts
+
+To support these updates the SDK will need to update the available methods on some transactions
+
+1. `CryptoUpdate` transactions will need support to add/remove and specify default `evmAddress`.
```
Comment (MiroslavGatsanoga) [inline HIP/hip-631.md:417]: We should add disable as an option here as well.

File: HIP/hip-631.md [hunk 15]
```diff
@@ -0,0 +1,555 @@
...
+### Alias to Virtual Account Migration
+
+To dissociate the alias concept from EVM public addresses and to reduce the overhead on users the network should perform a state migration for accounts and contracts
+
+#### Accounts
+
+1. An `evmAddress` → `accountId` map is created to resolve future virtual address references
+2. Traverse all accounts 
+3. If `account.key` is not a simple ECDSA key (`evmAddress`) skip
+4. If a simple `ECDSA` key based accounts
```
[the remark this answers is not in the document]
Comment (Nana-EC) [reply HIP/hip-631.md:371]: Updated it to highlight alias first, then all ECDSA accounts in migration

File: HIP/hip-631.md (hunk 14 above)
Comment (Nana-EC) [reply HIP/hip-631.md:417]: Added

File: HIP/hip-631.md (hunk 13 above)
Comment (Nana-EC) [reply HIP/hip-631.md:266]: Added

File: HIP/hip-631.md [hunk 16]
```diff
@@ -0,0 +1,555 @@
...
+
+## Abstract
+
+The Hedera Smart Contract Service (HSCS) supports the execution of EVM transactions in a fast and cheap manner, while striving for strict EVM equivalency. 
+Hedera also supports a rich set of "native" APIs (the HAPI or Hedera API). At times, Hedera and Ethereum have made different design decisions. 
+To support EVM equivalency on Hedera, this HIP proposes a solution to one of these design points. 
+
+In Hedera, accounts are persistent entities with keys, and those keys can rotate over time. 
+In Ethereum, keys and accounts are synonymous, and the only way to rotate keys is to rotate accounts. 
+This HIP proposes a new conept, that of "Virtual Addresses", which bridge the conceptual gap between Hedera's view of accounts, and Ethereum's.
```
[the remark this answers is not in the document]
Comment (Nana-EC) [reply HIP/hip-631.md:24]: updated

File: HIP/hip-631.md [hunk 17]
```diff
@@ -0,0 +1,555 @@
...
+
+For easy reference new and clarifying terms are listed below.
+
+- **ETH contracts**: An EVM contract currently deployed to Ethereum or other EVM compatible ledgers in which a maintainer does not have to change code in any way before deploying to Hedera. This code may include the use of precompile contract operations such as `ECRECOVER`.
+
+- **Hedera contracts**: Smart Contracts implemented with Hedera-specific extensions such as HTS integration.
+
+- **Hedera address**: Also known as the `long-zero` address (as noted in [HIP 583](https://hips.hedera.com/hip/hip-583)) account identifier, this is the result of encoding the `<shard>.<relam>.<num>` integer values into a 20 byte hex format to be conformant with `ECDSA` public addresses.
```
[the remark this answers is not in the document]
Comment (Nana-EC) [reply HIP/hip-631.md:32]: updated

File: HIP/hip-631.md (hunk 8 above)
Comment (Nana-EC) [reply HIP/hip-631.md:65]: updated

File: HIP/hip-631.md (hunk 9 above)
Comment (Nana-EC) [reply HIP/hip-631.md:67]: updated

File: HIP/hip-631.md (hunk 10 above)
Comment (Nana-EC) [reply HIP/hip-631.md:164]: updated

File: HIP/hip-631.md (hunk 11 above)
Comment (Nana-EC) [reply HIP/hip-631.md:341]: updated

File: HIP/hip-631.md (hunk 12 above)
Comment (Nana-EC) [reply HIP/hip-631.md:343]: updated

**Review by mgarbs — APPROVED** [review]

**Pull request merged**